### PR TITLE
Lock down types for SwitchCombinator, cast / throw in concrete idx case (GEN-486)

### PR DESCRIPTION
**Pull request by sritchie**
@littleredcomputer and I discovered that `SwitchCombinator` ALREADY casts / requires that all of the incoming generative functions have the same return type, thanks to `jnp.choose` (invoked when an index is not an `int`).

This behavior allowed the user to see different return types when they passed an `int` index vs an array.

This PR:
- removes this difference in behavior by _always_ calling `jnp.choose` (to benefit from its shape-and-type checking and errors), but then in the case of a concrete `int` index, selecting using `vs[idx]` instead of returning the result of `jnp.choose`.
- types `SwitchCombinator` as `GenerativeFunction[R | Sum[R]]` and pushes this through `mix` and `or_else`
- swaps out `Sum.maybe_none` calls throughout `SwitchCombinator` for `Sum.maybe` calls

### CODE REVIEW

File: src/genjax/_src/core/generative/functional_types.py [hunk 1]
```diff
@@ -182,20 +182,16 @@ def maybe(
...
+        return Sum[R](idx, vs).maybe_collapse()
 
     @classmethod
     def maybe_none(
         cls,
         idx: ArrayLike | Diff[Any],
         vs: list[R],
-    ):
-        possibles = []
+    ) -> "R | Mask[R] | Sum[R] | None":
```
Comment (sritchie) [inline src/genjax/_src/core/generative/functional_types.py:193]: @femtomc notice this wacky type... you can get a `Mask[R]` out if you pass a singleton list with a concrete index. That can't be intentional, right?

**Review by femtomc — APPROVED** [review]
